splink/truth.py

```diff
@@ -0,0 +1,444 @@
+from .settings import complete_settings_dict
+
+from functools import reduce
+from pyspark.sql import DataFrame
+
+altair_installed = True
+try:
+    import altair as alt
+except ImportError:
+    altair_installed = False
+
+try:
+    from pyspark.sql.dataframe import DataFrame
+    from pyspark.sql.session import SparkSession
+except ImportError:
+    DataFrame = None
+    SparkSession = None
+
+
+def _sql_gen_unique_id_keygen(table, uid_col1, uid_col2):
+
+    return f"""
+    case
+    when {table}.{uid_col1} > {table}.{uid_col2} then concat({table}.{uid_col2}, '-', {table}.{uid_col1})
+    else concat({table}.{uid_col1}, '-', {table}.{uid_col2})
+    end
+    """
+
+
+def _check_df_labels(df_labels, settings):
+
+    cols = df_labels.columns
+    colname = settings["unique_id_column_name"]
+
+    assert f"{colname}_l" in cols, f"{colname}_l should be a column in df_labels"
+    assert f"{colname}_r" in cols, f"{colname}_l should be a column in df_labels"
+    assert (
+        "clerical_match_score" in cols
+    ), f"clerical_match_score should be a column in df_labels"
+
+
+def _get_score_colname(settings):
+    score_colname = "match_probability"
+    for c in settings["comparison_columns"]:
+        if c["term_frequency_adjustments"] is True:
+            score_colname = "tf_adjusted_match_prob"
+    return score_colname
+
+
+def _join_labels_to_results(df_labels, df_e, settings, spark):
+
+    # df_labels is a dataframe like:
+    # | unique_id_l | unique_id_r | clerical_match_score |
+    # |:------------|:------------|---------------------:|
+    # | id1         | id2         |                  0.9 |
+    # | id1         | id3         |                  0.1 |
+    settings = complete_settings_dict(settings, None)
+
+    _check_df_labels(df_labels, settings)
+
+    uid_colname = settings["unique_id_column_name"]
+
+    # If settings has tf_afjustments, use tf_adjusted_match_prob else use match_probability
+    score_colname = _get_score_colname(settings)
+
+    # The join is trickier than it looks because there's no guarantee of which way around the two ids are
+    # it could be id1, id2 in df_labels and id2,id1 in df_e
+
+    uid_col_l = f"{uid_colname}_l"
+    uid_col_r = f"{uid_colname}_r"
+
+    df_labels.createOrReplaceTempView("df_labels")
+    df_e.createOrReplaceTempView("df_e")
+
+    sql = f"""
+    select
+
+    df_labels.{uid_col_l},
+    df_labels.{uid_col_r},
+    clerical_match_score,
+
+    case
+    when {score_colname} is null then 0
+    else {score_colname}
+    end as {score_colname},
+
+    case
+    when {score_colname} is null then false
+    else true
+    end as found_by_blocking
+
+
+    from df_labels
+    left join df_e
+    on {_sql_gen_unique_id_keygen('df_labels', uid_col_l, uid_col_r)}
+    = {_sql_gen_unique_id_keygen('df_e', uid_col_l, uid_col_r)}
+
+    """
+
+    return spark.sql(sql)
+
+
+def _categorise_scores_into_truth_cats(
```

Comment:
A suggestion: I would like more information (in comment format like in `_join_labels_to_results` ) on the kind of dataframe input expected and the kind of output dataframe expected. This will help with being called from other parts of the code (like the upcoming updated diagnostics.py)